perp/man/perpok.8
.\" perpok.8
.\" wcm, 2009.11.10 - 2011.02.01
.\" ===
.TH perpok 8 "March 2011" "perp-2.04" "persistent process supervision"
.SH NAME
perpok \- check a
.BR perpd (8)
service
.SH SYNOPSIS
.B perpok [\-hV] [\-b
.I basedir
.B ] [\-u
.I secs
.B ] [\-v]
.I sv
.SH DESCRIPTION
.B perpok
is used to check that the service
.I sv
is actively running under supervision of 
.BR perpd (8).
.PP
The argument
.I sv
should be a service definition directory supervised by
.BR perpd (8).
.B perpok
will look for
.I sv
in the directory given with the
.B \-b
option,
or as set in the
.B PERP_BASE
environmental variable,
or in the current directory if neither of the previous is given.
.PP
.B perpok
returns 0 to indicate success if the definition directory
.I sv
exists and is under the active supervision of
.BR perpd (8).
.PP
The
.B \-u
option may be given to specify a minimum uptime in
.IR secs .
In this case,
.B perpok
will further check that the main service itself is running,
is not resetting,
does not want down,
and that its uptime has been at least
.I secs
seconds.
If all these conditions are met,
.B perpok
returns 0 to indicate success.
.PP
If the conditions for success are not met,
.B perpok
exits non-zero.
.PP
.B perpok
is intended primarily as a utility for boolean testing in scripting environments.
See the
.BR perpstat (8)
and
.BR perpls (8)
utilities for more detailed status reporting.
.SH OPTIONS
.TP
.B \-b basedir
Base directory.
Sets the base directory containing the service definition
.IR sv .
If not set,
.B perpok
will use the value set in the variable PERP_BASE,
or the current directory if neither of these are defined.
On a normal system,
the base directory will be
.IR /etc/perp .
.TP
.B \-h
Help.
Print a brief usage message to stderr and exit.
.TP
.B \-u secs
Uptime.
Normally
.B perpok
checks only that the
.I sv
directory is active (exists and is sticky),
and that
.BR perpd (8)
is actively supervising it.
This option extends the checks performed by
.B perpok
to test that the main service itself is running,
is not resetting,
does not want down,
and that its uptime has been at least
.I secs
seconds.
Normally it is sensible to set
.I secs
to some small value, such as 2 or 3,
which is
just long enough to check that the service is not cycling on restarts.
.TP
.B \-V
Version.
Print the version number to stderr and exit.
.TP
.B \-v
Verbose.
Normally
.B perpok
exits silently,
as it is designed primarily for runscripts that might use it to perform simple dependency checks
prior to starting their own services.
This option causes
.B perpok
to print a message reporting either success or failure to stderr on exit.
.SH EXAMPLES
.B perpok
may be used as a primitive dependency tool within service runscripts:
.PP
.RS
.nf
#!/bin/sh
if test ${1} = "start" ; then
  if ! perpok -u3 foo ; then
    echo "dependency foo not running"
    exit 1
  fi
  # otherwise foo ok, continue to start service:
  exec whatever ...
fi
.fi
.RE
.PP
This example shows a runscript that uses
.B perpok
to require a dependency on another service named ``foo''.
If foo has not been running for at least 3 seconds,
the runscript exits immediately.
The
.BR perpd (8)
supervisor running this service will then continue trying to restart it at one second intervals.
.PP
Note that, when using
.B perpok
in a normal collection of services running under
.BR perpd (8),
PERP_BASE will always be properly predefined in the runscript environment.
.PP
.SH EXIT STATUS
.B perpok
exits with the following values:
.TP
0
Success.
The service
.I sv
is active and a
.BR perpd (8)
supervisor has been found running for it.
Additionally,
if the
.B \-u
option has been given,
the main service itself has been found running,
not resetting,
not wanting down,
and with an uptime of at least
.I secs
seconds.
.TP
1
Failure.
The service directory
.I sv
was found not to be active,
or a
.BR perpd (8)
supervisor was not found running for it,
or the additional constraints imposed for uptime checking
with the
.B \-u
option were not met.
.TP
100
Usage error.
For unknown options, missing arguments, or if
.I sv
does not name a directory.
Prints a brief diagnostic message to stderr on exit.
.TP
111
System error.
Unexpected failures during system calls, privilege and/or resource problems,
or configuration errors in the
.BR perpd (8)
system.
Prints a brief diagnostic message to stderr on exit.
.SH CAVEATS
For use as a dependency tool,
.B perpok
by itself is of course limited.
As illustrated in the example section above,
.B perpok
may find a dependency running okay at startup,
but there is no guarantee that it will continue to run for as long
as the dependent service requires it.
.PP
A collection of services running under
.BR perpd (8)
tends to be reliable,
so that the use of
.B perpok
for simple dependency checking at startup is often adequate.
Interrelated services may also use
.B perpok
in combination with runscript
.B reset
targets to implement a system of stricter dependency controls. 
.SH AUTHOR
Wayne Marshall, http://b0llix.net/perp/
.SH SEE ALSO
.nh
.BR perp_intro (8),
.BR perpboot (8),
.BR perpctl (8),
.BR perpd (8),
.BR perpetrate (5),
.BR perphup (8),
.BR perpls (8),
.BR perpstat (8),
.BR sissylog (8),
.BR tinylog (8)
.\" EOF perpok.8


CHANGES
# CHANGES
# perp release history and changelog
# wcm, 2010.02.17 - 2011.03.22
# ===

perp-2.04 (2011.03.22):
 * perphup:
   - added option -t, trigger perpd shutdown
 * perpboot:
   - waitpid() for deux[1] on SIGTERM
 * perp-setup:
   - added arg for ctldir
   - configure symlink /etc/perp/.control --> ctldir
   - configure rc.perp bootscript for runchoom(8)
 * runchoom:
   - new, utility for linux "oom killer" abatement

perp-2.03 (2011.03.11):
 * perp-setup:
   - in rc.perp boot script for perpd(8), setup default runlimit(8)
     configuration to hard limits for RLIMIT_NPROC, RLIMIT_NOFILE

perp-2.02 (2011.02.22):
 * perpd:
   - eliminated some spurious/redundant client connection logging

perp-2.01 (2011.02.11):
 * perp-setup:
   - runlimit added to .boot/rc.perp script
 * perpd:
   - added option -g, enable group permission on control socket
   - additional comments in LIMITS section of perpd(8) manual
 * runlimit:
   - added option -E, take resource limits from environment
   - added option -F, take resource limits from file
 * lasagna:
   - outvec/mk_outvec.c: portability fix for dietlibc

perp-2.00 (2011.02.02):
 * perpd:
   - coalesce supervisor (formerly perpetrate(8)) into scanner/controller
   - runs as a single process
   - all ipc with client applications via single domain socket 
   - ipc protocol version 2 (see PROTO_V2.txt)
 * perpetrate:
   - eliminate perpetrate(8) executable (subsumed by perpd(8))
   - perpetrate(5) manual page retained for runscript conventions
 * perpctl:
   - add commands [A | X], activate/deactivate service
 * libasagna:
   - nuscan.h, improve numeric scanning interface (supercedes nscan.h)
   - other iterative improvements incorporated from projects hdb and ezcdb
 
perp-0.00 (2010.01.12):
 - maiden public release

### EOF


perp/perpok.c
/* perpok.c
** perpok: check perpetrate supervisor
** wcm, 2009.11.09 - 2009.12.06
** ===
*/

/* standard library: */
#include <stdlib.h>

/* unix: */
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>

/* lasagna: */
#include "cstr.h"
#include "nextopt.h"
#include "nscan.h"
#include "sysstr.h"
#include "tain.h"
#include "uchar.h"

/* perp: */
#define  EPUTS_DEVOUT
#include "eputs.h"
#include "perp_common.h"
#include "perplib.h"

static const char *progname = NULL;
static const char prog_usage[] = " [-hV] [-b basedir] [-u secs] [-v] sv";


#define  report_fail(...) \
  {\
      if(verbose){\
          eputs(progname, ": ", svdir, ": fail: ", __VA_ARGS__);\
      }\
      die(1);\
  }

#define  report_ok(...) \
  {\
      if(verbose){\
          eputs(progname, ": ", svdir, ": ok: ",  __VA_ARGS__);\
      }\
      die(0);\
  }


int main(int argc, char *argv[])
{
   char           opt;
   nextopt_t      nopt = nextopt_INIT(argc, argv, ":hVb:u:v");
   int            verbose = 0;
   uint32_t       arg_upsecs = 0;
   const char    *nscan_end;
   const char    *basedir = NULL;
   const char    *svdir = NULL;
   struct stat    st;
   const char    *ctlpath;
   const uchar_t *binstat;
   perpstat_t     perpstat;
   uchar_t        flags;
   tain_t         now;

   progname = nextopt_progname(&nopt);
   while ((opt = nextopt(&nopt))) {
      char           optc[2] = { nopt.opt_got, '\0' };
      switch (opt) {
      case 'h':
         usage();
         die(0);
         break;
      case 'V':
         version();
         die(0);
         break;
      case 'b':
         basedir = nopt.opt_arg;
         break;
      case 'u':
         arg_upsecs = nscan_uint32(nopt.opt_arg, &nscan_end);
         if (*nscan_end != '\0') {
            fatal_usage("numeric argument required for option -", optc);
         }
         break;
      case 'v':
         ++verbose;
         break;
      case ':':
         fatal_usage("missing argument for option -", optc);
         break;
      case '?':
         if (nopt.opt_got != '?') {
            fatal_usage("invalid option -", optc);
         }
         /* else fallthrough: */
      default:
         die_usage();
         break;
      }
   }

   argc -= nopt.arg_ndx;
   argv += nopt.arg_ndx;

   svdir = argv[0];
   if (svdir == NULL) {
      fatal_usage("missing argument");
   }

   if (!basedir)
      basedir = getenv("PERP_BASE");
   if (!basedir)
      basedir = ".";

   if (chdir(basedir) != 0) {
      fatal_syserr("fail chdir() to ", basedir);
   }

   if (stat(svdir, &st) == -1) {
      fatal_syserr("fail stat() on service directory ", svdir);
   }

   if (!S_ISDIR(st.st_mode)) {
      fatal_usage("argument not a directory: ", svdir);
   }

   if (!(S_ISVTX & st.st_mode)) {
      report_fail("service directory not activated");
   }

   /* get path to service control directory (relative to cwd): */
   ctlpath = perp_ctlpath(&st);

   if (perp_ready(ctlpath) == -1) {
      report_fail("supervisor not running\n");
   }

   if (arg_upsecs == 0) {
      /* all done: */
      report_ok("supervisor running ok");
   }

   /*
    ** else: extended testing...
    */

   /* read binary-encoded status: */
   binstat = perp_binstat(ctlpath);
   if (binstat == NULL) {
      if (errno == EPROTO) {
         report_fail("bad status format found in ", STATUS_BIN);
      } else {
         report_fail("error reading status in ", STATUS_BIN, ": ",
                     sysstr_errno(errno));
      }
   }

   /* decode status: */
   perp_statload(&perpstat, binstat);

   if (perpstat.subsv[SUBSV_MAIN].pid == 0) {
      report_fail("service not running (pid is 0)");
   }

   flags = perpstat.subsv[SUBSV_MAIN].flags;
   if (subsv_isreset(flags)) {
      report_fail("service resetting");
   }
   if (subsv_iswant(flags)) {
      report_fail("service wants down");
   }

   /* test uptime compared to now: */
   tain_now(&now);
   if ((perp_uptime(&now, &perpstat.subsv[SUBSV_MAIN].when)) < arg_upsecs) {
      report_fail("service uptime not met");
   }

   /* else okiedokeyshmokey: */
   report_ok("service uptime ok");

   /* not reached: */
   return 0;
}


/* eof: perpok.c */
